internal/services/pam/pam_test.go

```
@@ -0,0 +1,606 @@
+		"Error when starting the session":                 {username: "NS_error", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			brokerCfgDir := t.TempDir()
+			brokerCfg := testutils.StartBusBrokerMock(t, brokerCfgDir)
+
+			client := startClient(t, brokerCfgDir)
+
+			if tc.brokerID == "" {
+				tc.brokerID = getGeneratedID(t, client, filepath.Base(brokerCfg))
+			} else if tc.brokerID == "empty" {
+				tc.brokerID = ""
+			}
+
+			sbRequest := &authd.SBRequest{
+				BrokerId: tc.brokerID,
+				Username: tc.username,
+			}
+			sbResp, err := client.SelectBroker(context.Background(), sbRequest)
+			if tc.wantErr {
+				require.Error(t, err, "SelectBroker should return an error, but did not")
+				return
+			}
+			require.NoError(t, err, "SelectBroker should not return an error, but did")
+			got := fmt.Sprintf("ID: %s\nEncryption Key: %s\n",
+				strings.ReplaceAll(sbResp.GetSessionId(), tc.brokerID, "BROKER_ID"),
+				sbResp.GetEncryptionKey())
+			want := testutils.LoadWithUpdateFromGolden(t, got)
+			require.Equal(t, want, got, "SelectBroker returned an unexpected response")
+		})
+	}
+}
+
+func TestGetAuthenticationModes(t *testing.T) {
+	t.Parallel()
+
+	required, optional := "required", "optional"
+	supportedEntries := "optional:chars,chars_password"
+	optionalWithBooleans := "optional:true,false"
+
+	normalLayout := &authd.UILayout{
+		Type:    "form",
+		Label:   &required,
+		Button:  &optional,
+		Wait:    &optionalWithBooleans,
+		Entry:   &supportedEntries,
+		Content: &optional,
+	}
+	invalidLayout := &authd.UILayout{
+		Type: "",
+	}
+
+	tests := map[string]struct {
+		brokerID  string
+		sessionID string
+		username  string
+		layout    *authd.UILayout
+
+		wantErr bool
+	}{
+		"Successfully get authentication modes": {},
+
+		"Error when sessionID is empty":           {sessionID: "empty", wantErr: true},
```

I think that all family tests after SelectBroker should have a "fail with not started session" test case (give an invalid session id basically), That ways, it’s clear that we only have valid calls if we selected a broker first to start a session.

---

Which is the `"Error when broker does not exist"` that exists in all family tests, no?